res/controllers/Hercules-DJControl-Inpulse-300-script.js

```diff
@@ -44,76 +60,239 @@ DJCi300.scratchShiftMultiplier = 4;
+    };
+    // Slicer beat (stores what beat slicer is on)
+    DJCi300.slicerBeat = {
+        1: 0,
+        2: 0
+    };
+
+    // Loop variables
+    // loopMode is on (equal to 1) if the current loop is created by the "Loop In" button
+    // and off (equal to 0) if it is created by the slicer buttons
+    DJCi300.loopMode = {
+        1: 0,
+        2: 0
     };
 
     // Turn On Vinyl buttons LED(one for each deck).
     midi.sendShortMsg(0x91, 0x03, 0x7F);
     midi.sendShortMsg(0x92, 0x03, 0x7F);
 
-    //Turn On Browser button LED
+    // Turn On Browser button LED
     midi.sendShortMsg(0x90, 0x04, 0x05);
 
-   //Softtakeover for Pitch fader
-   engine.softTakeover("[Channel1]", "rate", true);
-   engine.softTakeover("[Channel2]", "rate", true);
-   engine.softTakeoverIgnoreNextValue("[Channel1]", "rate");
-   engine.softTakeoverIgnoreNextValue("[Channel2]", "rate");
-
-   // Connect the VUMeters
-    engine.connectControl("[Channel1]", "vu_meter", "DJCi300.vuMeterUpdateDeck");
-	engine.getValue("[Channel1]", "vu_meter", "DJCi300.vuMeterUpdateDeck");
-    engine.connectControl("[Channel2]", "vu_meter", "DJCi300.vuMeterUpdateDeck");
-	engine.getValue("[Channel2]", "vu_meter", "DJCi300.vuMeterUpdateDeck");
-    engine.connectControl("[Main]", "vu_meter_left", "DJCi300.vuMeterUpdateMaster");
-    engine.connectControl("[Main]", "vu_meter_right", "DJCi300.vuMeterUpdateMaster");
-	engine.getValue("[Main]", "vu_meter_left", "DJCi300.vuMeterUpdateMaster");
-    engine.getValue("[Main]", "vu_meter_right", "DJCi300.vuMeterUpdateMaster");
+    // Turn On Toneplay LED
+    midi.sendShortMsg(0x96, 0x40, 0x7F);
+    midi.sendShortMsg(0x96, 0x48, 0x7F);
+    midi.sendShortMsg(0x97, 0x40, 0x7F);
+    midi.sendShortMsg(0x97, 0x48, 0x7F);
+
+    //Softtakeover for Pitch fader
+    engine.softTakeover("[Channel1]", "rate", true);
+    engine.softTakeover("[Channel2]", "rate", true);
+    engine.softTakeoverIgnoreNextValue("[Channel1]", "rate");
+    engine.softTakeoverIgnoreNextValue("[Channel2]", "rate");
+
+    // Connect the VUMeters
+    engine.makeConnection("[Channel1]", "VuMeter", DJCi300.vuMeterUpdateDeck);
+    engine.makeConnection("[Channel2]", "VuMeter", DJCi300.vuMeterUpdateDeck);
+    engine.makeConnection("[Master]", "VuMeterL", DJCi300.vuMeterUpdateMaster);
+    engine.makeConnection("[Master]", "VuMeterR", DJCi300.vuMeterUpdateMaster);
+
+    // Connect beatmatch LED functions
+    engine.makeConnection("[Channel1]", "bpm", DJCi300.updateBeatmatchTempoLED);
+    engine.makeConnection("[Channel2]", "bpm", DJCi300.updateBeatmatchTempoLED);
+    // We only connect channel 1 here because beatmatch is only enabled when both decks are connected anyway
+    // This saves resources
+    engine.makeConnection("[Channel1]", "beat_distance", DJCi300.updateBeatmatchAlignLED);
+    // We also want to update all beatmatch LEDs when a song is played
+    engine.makeConnection("[Channel1]", "play", DJCi300.updateBeatmatchAlignLED);
```

Comment:
does this need to be two separate connections to two functions? why not one function that does both things?

Reply:
I wanted to separate the beat align and tempo update functions. That way, if the tempo is synced but the beats aren't aligned, the program doesn't have to keep checking the tempo. In theory, this is computationally cheaper, although I don't know by how much. 

`updateBeatmatchTempoLED` is connected to the tempo of both decks, since if either deck changes BPM, it could cause the decks to desync.
`updateBeatmatchAlignLED` is only connected to the beat distance of one deck, since it just has to execute on every tick.

I suppose another implementation would be to create a giant beatmatch update function, but it would still have to be connected to `bpm` and `play` on both decks, as well as `beat_distance` on one deck, with the performance downside I mentioned earlier (unless I'm missing something).